src/Catalog/Dnx/DnxMaker.cs

```diff
@@ -260,6 +274,122 @@ private async Task<Uri> SaveNupkgAsync(Stream nupkgStream, Storage storage, stri
+                return;
+            }
+
+            var iconPath = PathUtility.StripLeadingDirectorySeparators(iconFilename);
+
+            var destinationRelativeUri = GetRelativeAddressIcon(packageId, normalizedPackageVersion);
+            var destinationUri = destinationStorage.ResolveUri(destinationRelativeUri);
+
+            if (disposeOfPackageStream)
+            {
+                using (var packageStream = await getPackageStream())
+                {
+                    await ExtractAndStoreIconAsync(packageStream, iconPath, destinationStorage, destinationUri, cancellationToken);
+                }
+            }
+            else
+            {
+                var packageStream = await getPackageStream();
+                await ExtractAndStoreIconAsync(packageStream, iconPath, destinationStorage, destinationUri, cancellationToken);
+            }
+        }
+
+        private async Task ExtractAndStoreIconAsync(
+            Stream packageStream,
+            string iconPath,
+            Storage destinationStorage,
+            Uri destinationUri,
+            CancellationToken cancellationToken)
+        {
+            using (var zipArchive = new ZipArchive(packageStream, ZipArchiveMode.Read, leaveOpen: true))
+            {
+                var iconEntry = zipArchive.Entries.FirstOrDefault(e => e.FullName.Equals(iconPath, StringComparison.InvariantCultureIgnoreCase));
+                if (iconEntry != null)
+                {
+                    using (var iconStream = iconEntry.Open())
+                    {
+                        var iconContent = new StreamStorageContent(iconStream, GetIconContentType(iconPath), DnxConstants.DefaultCacheControl);
+                        await destinationStorage.SaveAsync(destinationUri, iconContent, cancellationToken);
+                    }
+                }
+            }
+        }
+
+        private async Task<Stream> GetPackageStreamAsync(
+            IAzureStorage sourceStorage,
+            string packageId,
+            string normalizedPackageVersion,
+            CancellationToken cancellationToken)
+        {
+            var packageFileName = PackageUtility.GetPackageFileName(packageId, normalizedPackageVersion);
+            var sourceUri = sourceStorage.ResolveUri(packageFileName);
+            var packageSourceBlob = await sourceStorage.GetCloudBlockBlobReferenceAsync(sourceUri);
+            return await packageSourceBlob.GetStreamAsync(cancellationToken);
+        }
+
+        private string GetIconContentType(string iconPath)
+        {
+            if (iconPath.EndsWith(".png", StringComparison.InvariantCultureIgnoreCase))
+            {
+                return "image/png";
+            }
+            if (iconPath.EndsWith(".jpg", StringComparison.InvariantCultureIgnoreCase) || iconPath.EndsWith(".jpeg", StringComparison.InvariantCultureIgnoreCase))
+            {
+                return "image/jpeg";
+            }
+            return "image/jpeg";
```

Comment:
>image [](start = 20, length = 5)

Should we maybe throw here?

Reply:
I don't want package publishing to fail because of some icon issues. I'd rather have it published with wrong type and have client do its magic.

Reply:
> I don't want package publishing to fail because of some icon issues. 

I agree, but we should be able to catch this specific problem in the Gallery when the package is uploaded, right?

> I'd rather have it published with wrong type and have client do its magic.

Will the client be able to properly handle an image with the wrong mime type?

Reply:
Is there a mimetype that means "image but not sure what kind"? JPEG is misleading.

We should definitely have telemetry when this event happens.
Perhaps we should even have a metric that logs the file types used by images?
At a bare minimum, I think we should log a warning when it's not as JPEG or PNG.

Also, add a comment to this code so we know what's going on.

(nit: add a line between this and the `}`)

Reply:
Gallery is supposed to validate everything related to the embedded icon. Client already handles invalid mime types since there are 0 guarantees regarding where iconUrl points to.

Reply:
I did not find a mime type that means "some sort of image". But I ran some tests: browsers seem to ignore the specific type of an image (i.e. if server returns jpeg and says it's `image/png` browsers display it correctly). VS client does something similar.

Reply:
Interesting. Did the image have an extension in your tests?

Reply:
We need a comment about the arbitrary mime-type selection. What if you just leave the mime-type blank in this case?

Reply:
Spoke offline. Maybe we should drop the mime-type determination till we know what gallery is doing here in terms of validation.

Reply:
Removed for now.